### PR TITLE
LRU improvement's part 1

### PR DESCRIPTION
**Pull request by Alan-Jowett**
## Description

Merge the storage for LRU data with value data stored in hash table.

## Testing

CI/CD

## Documentation

No.

Resolves: #557

### CODE REVIEW

**Comment by Alan-Jowett** [comment]
```
Merged:
test_bpf_map_lookup_elem_read<BPF_MAP_TYPE_LRU_HASH>,1,8594
test_bpf_map_lookup_elem_read<BPF_MAP_TYPE_LRU_HASH>,0,8582
test_bpf_map_lookup_elem_write<BPF_MAP_TYPE_LRU_HASH>,1,8567
test_bpf_map_lookup_elem_write<BPF_MAP_TYPE_LRU_HASH>,0,8569
test_bpf_map_update_elem<BPF_MAP_TYPE_LRU_HASH>,1,9574
test_bpf_map_update_elem<BPF_MAP_TYPE_LRU_HASH>,0,9340
test_bpf_map_update_lru_elem<BPF_MAP_TYPE_LRU_HASH>,1,30505
test_bpf_map_update_lru_elem<BPF_MAP_TYPE_LRU_HASH>,0,21399

Baseline:
test_bpf_map_lookup_elem_read<BPF_MAP_TYPE_LRU_HASH>,1,9417
test_bpf_map_lookup_elem_read<BPF_MAP_TYPE_LRU_HASH>,0,9442
test_bpf_map_lookup_elem_write<BPF_MAP_TYPE_LRU_HASH>,1,9304
test_bpf_map_lookup_elem_write<BPF_MAP_TYPE_LRU_HASH>,0,9268
test_bpf_map_update_elem<BPF_MAP_TYPE_LRU_HASH>,1,10195
test_bpf_map_update_elem<BPF_MAP_TYPE_LRU_HASH>,0,10114
test_bpf_map_update_lru_elem<BPF_MAP_TYPE_LRU_HASH>,1,32905
test_bpf_map_update_lru_elem<BPF_MAP_TYPE_LRU_HASH>,0,17797
```

File: libs/execution_context/ebpf_maps.c [hunk 1]
```diff
@@ -921,12 +917,14 @@ _reap_oldest_map_entry(_In_ ebpf_core_map_t* map)
 
     // Walk through all the keys and values and find the oldest one.
     for (;;) {
+        uint64_t* key_age = NULL;
         result = ebpf_hash_table_next_key_pointer_and_value(
-            lru_map->key_history, previous_key, &next_key, (uint8_t**)&key_age);
+            (ebpf_hash_table_t*)lru_map->core_map.data, previous_key, &next_key, (uint8_t**)&value);
         if (result != EBPF_SUCCESS) {
```
Comment (dv-msft) [inline libs/execution_context/ebpf_maps.c:923]: should we add a check before the loop to see if the LRU is empty or partially filled and return 'True' if so?  reaping is really a 'no-op' in such cases so why bother proceeding?  Also, the error returned from within the loop would then become more meaningful.

To be clear, the current code calls this function only if the LRU is full.  But newer code might find some 'creative' uses for this function :-)

This is just a thought.  I'm fine with this implementation.

**Review by dv-msft — DISMISSED** [review]
Approved with a suggestion.

**Comment by Alan-Jowett** [comment]
Address sanitizer found an issue. We are incorrectly copying value size + supplemental value into the value field.

**Comment by Alan-Jowett** [comment]
Moving to draft until I fix the address sanitizer issue.

File: libs/execution_context/ebpf_maps.c [hunk 2]
```diff
@@ -705,6 +705,11 @@ _create_hash_map_internal(
         goto Done;
     }
 
+    retval = ebpf_safe_size_t_add(map_definition->value_size, supplemental_value_size, &actual_value_size);
```
Comment (dv-msft) [inline libs/execution_context/ebpf_maps.c:708]: If LRU updates are in a 'hot' path, would we end up paying 'un-aligned access' performance penalties if the 'key-age' value location is not 64-bit aligned?  Maybe we could round up the ```map_definition->value_size``` to a 64-bit boundary, thereby ensuring that the 'key-age' location is 64-bit aligned as well.

File: libs/execution_context/ebpf_maps.c (hunk 1 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c:918]: This block of code is going to be replaced in the next PR. In addition to what you pointed out, there is also the issue that after reaping one entry, the insert can still fail if another thread inserts before this one can.

File: libs/execution_context/ebpf_maps.c (hunk 2 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_maps.c]: Fixed.

**Review by gtrevi — APPROVED** [review]

**Comment by dthaler** [comment]
FYI, see https://patchwork.kernel.org/project/netdevbpf/patch/20221103204330.3257910-1-joe@isovalent.com/ for some documentation about LRU maps.

File: libs/execution_context/ebpf_maps.c [hunk 3]
```diff
@@ -31,9 +31,7 @@ typedef struct _ebpf_core_object_map
 typedef struct _ebpf_core_lru_map
 {
     ebpf_core_map_t core_map;
-    // https://github.com/microsoft/ebpf-for-windows/issues/557
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:34]: The PR description doesn't reference issue 557, should it?

File: libs/platform/ebpf_hash_table.c [hunk 4]
```diff
@@ -2,6 +2,7 @@
 // SPDX-License-Identifier: MIT
 
 #include "ebpf_platform.h"
+#include "ebpf_epoch.h"
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:5]: Alphabetize

File: libs/execution_context/ebpf_maps.c [hunk 5]
```diff
@@ -690,13 +688,13 @@ static ebpf_result_t
 _create_hash_map_internal(
     size_t map_struct_size,
     _In_ const ebpf_map_definition_in_memory_t* map_definition,
+    _In_ const size_t supplemental_value_size,
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:691]:
```suggestion
    const size_t supplemental_value_size,
```
Don't use `_In_` with value types, only pointers.  Also probably don't need const here either.

File: libs/execution_context/ebpf_maps.c [hunk 6]
```diff
@@ -937,13 +935,9 @@ _reap_oldest_map_entry(_In_ ebpf_core_map_t* map)
     // If we reached the end of the keys, delete the oldest one found.
     if (result == EBPF_NO_MORE_KEYS && oldest_key != NULL) {
         ebpf_result_t delete_result = ebpf_hash_table_delete((ebpf_hash_table_t*)lru_map->core_map.data, oldest_key);
-        // See issue #557 for why the delete can fail.
-        // https://github.com/microsoft/ebpf-for-windows/issues/557
-        // This can result in map having more entries than max_entries.
```
Comment (dthaler) [inline libs/execution_context/ebpf_maps.c:942]: Why can the delete still fail?  Assuming it can fail, add a comment explaining.

**Review by dthaler — APPROVED** [review]
Ok now except needs a rebase to resolve conflicts

**Review by dv-msft — APPROVED** [review]

**Pull request merged**